### ISSUE TITLE
Provide support or tooling to help generate an ECS template

### ISSUE THREAD

**Issue by MarkSettleES** [elastic/ecs#324]
During our [Intro to ECS webinar](https://www.elastic.co/webinars/introducing-the-elastic-common-schema) we received questions about how an organization that plans to use some custom fields in their implementation of ECS can keep track of what's standard and what they've changed. We shared best practices for things like naming. Toward the end, an attendee (who I'll keep anonymous, but who should know that we greatly appreciate his idea) suggested the following:

>Having a process to allow the customizations to be separate from the product would be good. Maybe an “extra.yml” file or something so that you don’t have to touch the base stuff at all.

Doing this would make customizing and maintaining ECS easier, enabling broader adoption and reducing support issues.

**Comment by webmat** [comment]
Thanks @MarkSettleES for recording this request here. I totally agree we need this.

I'm not sure what the best delivery of the feature is, though

* a script in ECS repo would be a good approach to prototype this quickly
* support for this kind of workflow could be added to the [Kibana index management section](http://localhost:5601/app/kibana#/management/elasticsearch/index_management/indices). There's no "create index" button there yet, but there could be, and it could optionally support this concept of crafting one's index template
* put up a documentation page explaining how people can achieve this by creating two concurrent index templates directly in Elasticsearch applying to the same index pattern. There's a setting we can use to determine in which order the merge is done.

**Comment by webmat** [comment]
If you don't mind, I changed the title of the issue. I really like this idea.

**Comment by webmat** [comment]
Pointer for later, but the new file generator (introduced in #336) could likely be improved to accept a second set of field definitions, like what ECS has in `schemas/*`; and output the user's custom index template.

It would need a more polished CLI experience, of course :-)

**Comment by vbohata** [comment]
Do you plan to support aliases? We would like to keep our current field naming while mapping fields to ECS by aliases whenever possible. It would be very useful to define custom mapping override like "map myfield.group_name to group.name using alias".

**Comment by webmat** [comment]
Yes, field aliases have been introduced in Elasticsearch recently.

The Beats migration from 6.7 to 7.0 will use them quite a bit. You could also leverage aliases to get the same effect.

There's one caveat to using field aliases. They only add entries to the index mapping. This means searching/aggregating on the alias will work as expected. However the source of the document is not modified. So the document will only contain the "real" field, not the alias in the JSON. This will affect any direct usage of the documents, such as API calls or even displaying them in a Kibana saved search.

**Comment by webmat** [comment]
This tool should support two kinds of customizations:

* Start with the ECS fields, and augment based on the user's custom field definitions to produce the resulting template.
* Conversely, it should also be possible to skip parts of ECS, to control the size of the resulting template/mapping. If a user is creating a custom index for a very specific type of events, they should be able to skip the field sets they will never use in that index.

**Comment by vbohata** [comment]
Currently I am creating some script for our field names for generation our templates, ... with support to map our fields to ECS (multiple versions depending to the document received). I found following things and features of Elasticsearch would help very much:
- Use one ECS dynamic mapping template shared for all defined indexes. This requires merging of dynamic template mapping (has been removed from ES).
- Convert beats templates to dynamic mapping (again, has to be merging support from ES with support of using aliases in dynamic mapping and support of merging default_fields).

This would make everything more clean. ECS could be defined once in one template, beats could have one template, our fields could be directly aliased to ECS fields without the need to define ECS fields in our template ... also the Kibana would show the really used fields in its suggestions instead of all static fields (this is btw very very confusing for our users) ... and of course it would reduce the ES master node resources needed to handle such big templates.
